Source/JavaScriptCore/runtime/ImportMap.cpp

```diff
@@ -254,18 +284,189 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
 }
 
 String ImportMap::integrityForURL(const URL& url) const
 {
     return url.isNull() ? String() : m_integrity.get(url);
 }
 
+// https://html.spec.whatwg.org/C/#merge-existing-and-new-import-maps
+void ImportMap::mergeExistingAndNewImportMaps(Ref<ImportMap>&& newImportMap, const ImportMap::Reporter& reporter)
+{
+    // 1. Let newImportMapScopes be a deep copy of newImportMap's scopes.
+    // 2. Let newImportMapImports be a deep copy of newImportMap's imports.
+    //
+    // Instead of copying we have moved the newImportMap here and are performing
+    // the algorithm's mutations directly on them. That's fine because the move
+    // guarantees that no one will use this map for anything else.
+    ImportMap::ScopesMap& newImportMapScopes = newImportMap->m_scopesMap;
+    ImportMap::SpecifierMap& newImportMapImports = newImportMap->m_imports;
+    ImportMap::IntegrityMap& newImportMapIntegrity = newImportMap->m_integrity;
+
+    // 3. For each scopePrefix → scopeImports of newImportMapScopes:
+    for (auto& scope : newImportMapScopes) {
+        ImportMap::SpecifierMap& scopeImports = scope.value;
+        // 3.1. For each pair of global's resolved module set:
+        //
+        // 3.1.1. If pair's referring script does not start with scopePrefix,
+        // continue.
+        //
+        // 3.1.2. For each specifier → url of scopeImports:
+        //
+        // 3.1.2.1. If pair's specifier starts with specifier, then:
+        //
+        //
+        // We are using a different algorithm here, where instead of a resolved
+        // module set, we have a scoped resolved module map. The map's keys are
+        // scope prefixes, and its values are a set of specifier prefixes that
+        // already exist in that scope. We grab the set of specifier prefixes using
+        // the current scope and then iterate over the scope's imports, removing any
+        // specifiers whose prefix is in the set.
+        auto iter = m_scopedResolvedModuleMap.find(AtomString(scope.key.string()));
+        if (iter != m_scopedResolvedModuleMap.end()) {
+            auto& currentResolvedSet = iter->value;
+            Vector<AtomString> specifiersToRemove;
+            for (auto& specifier : scopeImports.keys()) {
+                if (currentResolvedSet.find(specifier) != currentResolvedSet.end())
+                    specifiersToRemove.append(specifier);
+            }
+            for (auto& specifier : specifiersToRemove) {
+                // 3.1.2.1.1. The user agent may report the removed rule as a warning to
+                // the developer console.
+                reporter.reportWarning(makeString("An import map scope rule for specifier '"_s, specifier, "' was removed, as it conflicted with already resolved module specifiers."_s));
+                // 3.1.2.1.2. Remove scopeImports[specifier].
+                scopeImports.remove(specifier);
+            }
```

Comment:
Use `scopeImports.removeIf`. It can make these handling simpler.
So then we do not need `specifiersToRemove`.